File: kcmjs/cui/menu.js
/**
 * Menu widgets for CUI framework
 */
import { Component } from '../core/component.js';
import { Style } from '../core/renderer.js';

export class MenuItem {
    constructor(label, callback, key = null) {
        this.label = label;
        this.callback = callback;
        this.key = key;
    }
}

export class Menu extends Component {
    constructor(rect, { title = '', items = [] } = {}) {
        super(rect);
        this.title = title;
        this.items = items;
        this.selectedIndex = 0;
        this.style = new Style({ fg: 'white' });
        this.selectedStyle = new Style({ fg: 'black', bg: 'cyan', bold: true });
    }

    render(renderer) {
        if (!this.visible) return;

        // Border
        const tl = '┌', tr = '┐', bl = '└', br = '┘', h = '─', v = '│';
        
        // Top border with title
        renderer.moveCursor(this.rect.x, this.rect.y);
        let top = tl + h.repeat(this.rect.width - 2) + tr;
        if (this.title) {
            const titleText = ` ${this.title} `;
            const pos = Math.floor((this.rect.width - titleText.length) / 2);
            top = top.substring(0, pos) + titleText + top.substring(pos + titleText.length);
        }
        renderer.buffer.push(renderer.styleText(top, this.style));

        // Menu items
        let y = this.rect.y + 1;
        for (let i = 0; i < this.items.length; i++) {
            const item = this.items[i];
            const isSelected = i === this.selectedIndex;
            const style = isSelected ? this.selectedStyle : this.style;
            
            renderer.moveCursor(this.rect.x, y);
            renderer.buffer.push(renderer.styleText(v, this.style));
            
            let itemText = ' ';
            if (item.key) {
                itemText += `[${item.key}] `;
            }
            itemText += item.label;
            itemText = itemText.padEnd(this.rect.width - 2);
            
            renderer.buffer.push(renderer.styleText(itemText, style));
            renderer.buffer.push(renderer.styleText(v, this.style));
            y++;
        }

        // Fill empty space
        while (y < this.rect.y + this.rect.height - 1) {
            renderer.moveCursor(this.rect.x, y);
            renderer.buffer.push(renderer.styleText(v + ' '.repeat(this.rect.width - 2) + v, this.style));
            y++;
        }

        // Bottom border
        renderer.moveCursor(this.rect.x, this.rect.y + this.rect.height - 1);
        renderer.buffer.push(renderer.styleText(bl + h.repeat(this.rect.width - 2) + br, this.style));
    }

    handleKey(key) {
        // Arrow keys
        if (key === '\x1b[A') { // Up
            this.selectedIndex = Math.max(0, this.selectedIndex - 1);
            return true;
        } else if (key === '\x1b[B') { // Down
            this.selectedIndex = Math.min(this.items.length - 1, this.selectedIndex + 1);
            return true;
        } else if (key === '\r' || key === '\n') { // Enter
            const item = this.items[this.selectedIndex];
            if (item && item.callback) {
                item.callback();
            }
            return true;
        }

        // Hotkeys
        for (let i = 0; i < this.items.length; i++) {
            const item = this.items[i];
            if (item.key && key === item.key) {
                this.selectedIndex = i;
                if (item.callback) {
                    item.callback();
                }
                return true;
            }
        }

        return false;
    }
}

export class SelectList extends Component {
    constructor(rect, { items = [], multiSelect = false } = {}) {
        super(rect);
        this.items = items;
        this.multiSelect = multiSelect;
        this.selectedIndex = 0;
        this.selectedItems = new Set();
        this.style = new Style({ fg: 'white' });
        this.selectedStyle = new Style({ fg: 'black', bg: 'cyan' });
    }

    render(renderer) {
        if (!this.visible) return;

        let y = this.rect.y;
        for (let i = 0; i < Math.min(this.items.length, this.rect.height); i++) {
            const item = this.items[i];
            const isSelected = i === this.selectedIndex;
            const isChecked = this.selectedItems.has(i);
            const style = isSelected ? this.selectedStyle : this.style;
            
            renderer.moveCursor(this.rect.x, y);
            
            let text = '';
            if (this.multiSelect) {
                text += isChecked ? '[✓] ' : '[ ] ';
            }
            text += item;
            text = text.substring(0, this.rect.width);
            
            renderer.buffer.push(renderer.styleText(text, style));
            y++;
        }
    }

    handleKey(key) {
        if (key === '\x1b[A') { // Up
            this.selectedIndex = Math.max(0, this.selectedIndex - 1);
            return true;
        } else if (key === '\x1b[B') { // Down
            this.selectedIndex = Math.min(this.items.length - 1, this.selectedIndex + 1);
            return true;
        } else if (key === ' ' && this.multiSelect) { // Space
            if (this.selectedItems.has(this.selectedIndex)) {
                this.selectedItems.delete(this.selectedIndex);
            } else {
                this.selectedItems.add(this.selectedIndex);
            }
            return true;
        }
        return false;
    }
}

File: kcmpy/cui/menu.py
"""Interactive menu widgets."""
from ..core.component import Component, Rect
from ..core.renderer import Style
from ..events.event_system import Event, EventType
from typing import List, Callable, Optional
import platform


class MenuItem:
    """Single menu item."""
    
    def __init__(self, label: str, callback: Optional[Callable] = None, key: str = None):
        self.label = label
        self.callback = callback
        self.key = key  # Keyboard shortcut


class Menu(Component):
    """Interactive menu with keyboard navigation."""
    
    def __init__(self, rect: Rect, title: str = "", items: List[MenuItem] = None):
        super().__init__(rect)
        self.title = title
        self.items = items or []
        self.selected_index = 0
        self.style_normal = Style(fg='white')
        self.style_selected = Style(fg='black', bg='bright_cyan', bold=True)
        self.style_title = Style(fg='bright_yellow', bold=True)
        self.style_border = Style(fg='bright_blue')
        self.style_key = Style(fg='bright_green')
        
        # Use ASCII for Windows
        self.is_windows = platform.system() == 'Windows'
    
    def add_item(self, item: MenuItem):
        """Add menu item."""
        self.items.append(item)
    
    def handle_event(self, event: Event) -> bool:
        """Handle keyboard events."""
        if event.type != EventType.KEY_PRESS:
            return False
        
        key = event.data.get('key', '')
        
        # Arrow navigation
        if key == '\x1b[A':  # Up arrow
            self.selected_index = (self.selected_index - 1) % len(self.items)
            return True
        elif key == '\x1b[B':  # Down arrow
            self.selected_index = (self.selected_index + 1) % len(self.items)
            return True
        elif key == '\r' or key == '\n':  # Enter
            if self.items[self.selected_index].callback:
                self.items[self.selected_index].callback()
            return True
        else:
            # Check for keyboard shortcuts
            for i, item in enumerate(self.items):
                if item.key and key == item.key:
                    self.selected_index = i
                    if item.callback:
                        item.callback()
                    return True
        
        return False
    
    def render(self, renderer):
        """Render menu."""
        if not self.visible:
            return
        
        y = self.rect.y
        
        # Title with border
        if self.title:
            renderer.move_cursor(self.rect.x, y)
            
            if self.is_windows:
                # ASCII border for Windows
                title_line = f"+== {self.title} ==+"
            else:
                # Unicode border
                title_line = f"╔══ {self.title} ══╗"
            
            # Center and pad
            padding = (self.rect.width - len(title_line)) // 2
            line = ' ' * padding + title_line
            renderer.buffer.append(renderer.style_text(line, self.style_title))
            y += 1
        
        # Menu items
        for i, item in enumerate(self.items):
            renderer.move_cursor(self.rect.x, y)
            
            # Selection indicator
            if self.is_windows:
                indicator = "> " if i == self.selected_index else "  "
            else:
                indicator = "▶ " if i == self.selected_index else "  "
            
            # Keyboard shortcut
            key_hint = f"[{item.key}] " if item.key else ""
            
            # Full line
            line = f"{indicator}{key_hint}{item.label}"
            
            # Pad to width
            line = line.ljust(self.rect.width)
            
            # Apply style
            style = self.style_selected if i == self.selected_index else self.style_normal
            
            if item.key and i != self.selected_index:
                # Highlight key separately
                key_styled = renderer.style_text(f"[{item.key}]", self.style_key)
                line_text = f"{indicator}{key_styled} {item.label}"
                padding_needed = self.rect.width - len(f"{indicator}[{item.key}] {item.label}")
                line_text += ' ' * padding_needed
                renderer.buffer.append(line_text)
            else:
                renderer.buffer.append(renderer.style_text(line, style))
            
            y += 1
        
        # Bottom border
        if self.title:
            renderer.move_cursor(self.rect.x, y)
            
            if self.is_windows:
                bottom = "+" + "=" * (self.rect.width - 2) + "+"
            else:
                bottom = "╚" + "═" * (self.rect.width - 2) + "╝"
            
            padding = (self.rect.width - len(bottom)) // 2
            line = ' ' * padding + bottom
            renderer.buffer.append(renderer.style_text(line, self.style_border))


class SelectList(Component):
    """Simple selection list without borders."""
    
    def __init__(self, rect: Rect, items: List[str] = None):
        super().__init__(rect)
        self.items = items or []
        self.selected_index = 0
        self.style_normal = Style(fg='white')
        self.style_selected = Style(fg='black', bg='bright_white', bold=True)
        self.is_windows = platform.system() == 'Windows'
    
    def handle_event(self, event: Event) -> bool:
        """Handle keyboard events."""
        if event.type != EventType.KEY_PRESS:
            return False
        
        key = event.data.get('key', '')
        
        if key == '\x1b[A':  # Up
            self.selected_index = (self.selected_index - 1) % len(self.items)
            return True
        elif key == '\x1b[B':  # Down
            self.selected_index = (self.selected_index + 1) % len(self.items)
            return True
        
        return False
    
    def render(self, renderer):
        """Render selection list."""
        if not self.visible:
            return
        
        for i, item in enumerate(self.items[:self.rect.height]):
            renderer.move_cursor(self.rect.x, self.rect.y + i)
            
            if self.is_windows:
                prefix = "> " if i == self.selected_index else "  "
            else:
                prefix = "▶ " if i == self.selected_index else "  "
            
            line = f"{prefix}{item}".ljust(self.rect.width)
            
            style = self.style_selected if i == self.selected_index else self.style_normal
            renderer.buffer.append(renderer.style_text(line, style))
